crypto-ffi/src/wasm/mod.rs

```diff
@@ -78,96 +78,195 @@ pub extern "C" fn __stack_chk_fail() {
+pub enum ProteusError {
+    #[error("The requested session was not found")]
+    SessionNotFound,
+    #[error("We already decrypted this message once")]
+    DuplicateMessage,
+    #[error("The remote identity has changed")]
+    RemoteIdentityChanged,
+    #[error("Another Proteus error occurred but the details are probably irrelevant to clients")]
+    Other(Option<u16>),
+}
+
+#[cfg(feature = "proteus")]
+impl From<core_crypto::ProteusError> for ProteusError {
+    fn from(value: core_crypto::ProteusError) -> Self {
+        type SessionError = proteus_wasm::session::Error<core_crypto_keystore::CryptoKeystoreError>;
+        match value {
+            core_crypto::ProteusError::ProteusSessionError(SessionError::InternalError(
+                proteus_wasm::internal::types::InternalError::NoSessionForTag,
+            )) => Self::SessionNotFound,
+            core_crypto::ProteusError::ProteusSessionError(SessionError::DuplicateMessage) => Self::DuplicateMessage,
+            core_crypto::ProteusError::ProteusSessionError(SessionError::RemoteIdentityChanged) => {
+                Self::RemoteIdentityChanged
+            }
+            _ => Self::Other(value.error_code()),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error, strum::AsRefStr)]
+pub(crate) enum InternalError {
     #[error(transparent)]
-    CryptoError(#[from] CryptoError),
+    MlsError(#[from] MlsError),
+    #[cfg(feature = "proteus")]
     #[error(transparent)]
-    E2eError(#[from] E2eIdentityError),
+    ProteusError(#[from] ProteusError),
+    #[error("End to end identity error")]
+    E2eiError,
     #[error(transparent)]
     SerializationError(#[from] serde_wasm_bindgen::Error),
-    #[error("Failed lifting an enum")]
-    EnumError,
+    #[error("Unknown cyphersuite identifier")]
+    UnknownCyphersuite,
     #[error("Transaction rolled back. Last proteus error code: {proteus_error_code:?}. Uncaught JsError: {uncaught_error:?}")]
     TransactionFailed {
         uncaught_error: JsValue,
-        proteus_error_code: u32,
+        proteus_error_code: Option<u16>,
     },
 }
 
-#[derive(Debug, Clone, serde::Serialize)]
-#[serde(rename_all = "camelCase")]
-struct CoreCryptoJsRichError {
-    error_name: String,
-    message: String,
-    rust_stack_trace: String,
-    proteus_error_code: u32,
-}
+// This implementation is intended to be temporary; we're going to be completely restructuring the way we handle
+// errors in `core-crypto` soon. We can replace this with better error patterns when we do.
+//
+// Certain error mappings could apply to both MLS and Proteus. In all such cases, we map them to the MLS variant.
+// When we redesign the errors in `core-crypto`, these ambiguities should disappear anyway.
+impl From<CryptoError> for InternalError {
+    fn from(value: CryptoError) -> Self {
+        #[cfg(feature = "proteus")]
+        if value.proteus_error_code().is_some() {
+            return ProteusError::Other(value.proteus_error_code()).into();
```

Comment:
Hmm so right we only return `ProteusError:Other` but we could map on the error code code see `ProteusErrorKind`

Reply:
Good point! Made that change.